### PR TITLE
key-val store

### LINKED ISSUE #97: Feature: key-val store
Implement Veraison storage layer.
Initially this is a simple key-val store.
There must exist an interface that decouples the calling code from the specific DB technology used to implement it.

### PR DESCRIPTION
**Pull request by thomas-fossati**
interface and two implementations:
* generic SQL backend
* in-memory

Fixes #97

### CODE REVIEW

File: kvstore/README.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+```
+
+where:
+
+* `scheme` encodes the attestation format (e.g., "psa", "tcg-dice",
+"tpm-enacttrust", "open-dice", "tcg-tpm", etc.)
+* `authority` encodes the tenant
+* `path-absolute` encodes the parts of the key, identified positionally.  Missing optional parts are encoded as empty path segments.
+
+Attestation techology specific code (i.e., plugins) must provide their own syntesis functions.
```
Comment (yogeshbdeshpande) [inline kvstore/README.md:58]:
```suggestion
Attestation technology specific code (i.e., plugins) must provide their own synthesis functions.
```

File: kvstore/kvstore.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// Copyright 2021 Contributors to the Veraison project.
+// SPDX-License-Identifier: Apache-2.0
+package kvstore
+
+type KVStore interface {
+	Init(cfg Config) error
+	Term() error
```
Comment (yogeshbdeshpande) [inline kvstore/kvstore.go:7]:
```suggestion
	Close() error
```

File: kvstore/memory_test.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+		"store-type": "some-random-type",
+	}
+
+	expectedErr := `invalid "store-type": unknown type "some-random-type"`
+
+	err := s.Init(cfg)
+	assert.EqualError(t, err, expectedErr)
+}
+
+func TestMemoryStore_Init_bad_store_type(t *testing.T) {
```
Comment (yogeshbdeshpande) [inline kvstore/memory_test.go:58]: Tests on Line 32, 45 and 58 can be potentially combined into a single table driven test.

Comment (thomas-fossati) [reply kvstore/memory_test.go:58]: True, and others as well in this same lot could be made table-driven, I think.  Question: is it necessary?

File: kvstore/kvstore.go (hunk 2 above)
Comment (thomas-fossati) [reply kvstore/kvstore.go]: done in 0417174

File: kvstore/sql_test.go [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+	e.WithArgs(testKey)
+	e.WillReturnRows()
+
+	err = s.Del(testKey)
+	assert.NoError(t, err)
+
+	if err := mock.ExpectationsWereMet(); err != nil {
+		t.Errorf("unfulfilled expectations: %s", err)
+	}
+}
```
Comment (yogeshbdeshpande) [inline kvstore/sql_test.go:298]: I do not see any test which tests TestSQLStore_Init_Close_cycle_ok ?
I see close() executed naturally, I understand, but where is s.Init(cfg) +ive case?

**Review by yogeshbdeshpande — COMMENTED** [review]
Have completed my review!

File: kvstore/memory_test.go (hunk 3 above)
Comment (yogeshbdeshpande) [reply kvstore/memory_test.go:58]: Well, only benefit I see is the code compactness and logical coherence! Reader sees all negative tests in one place.

I leave this up to you on your best judgement!

File: kvstore/sql_test.go (hunk 4 above)
Comment (thomas-fossati) [reply kvstore/sql_test.go:298]: oops, I forgot.  Will add.

File: kvstore/memory_test.go (hunk 3 above)
Comment (thomas-fossati) [reply kvstore/memory_test.go:58]: On compactness I agree, logical coherence not so much.  Anyway, if you really prefer this way, I'll do it.

File: kvstore/sql_test.go (hunk 4 above)
Comment (thomas-fossati) [reply kvstore/sql_test.go:314]: Fixed in 0dfbb0e

File: kvstore/memory_test.go (hunk 3 above)
Comment (thomas-fossati) [reply kvstore/memory_test.go]: Addressed in 0dfbb0e

File: kvstore/sql_test.go [hunk 5]
```diff
@@ -71,6 +71,22 @@ func TestSQL_Init_db_open_unknown_driver_postgres(t *testing.T) {
 	assert.EqualError(t, err, expectedErr)
 }
 
+func TestSQL_Init_Close_cycle_ok(t *testing.T) {
+	s := Memory{}
```
Comment (thomas-fossati) [inline kvstore/sql_test.go:75]: arg!

Comment (thomas-fossati) [reply kvstore/sql_test.go]: it turns out we cannot run this test because it needs a SQL driver to work (see README.md)

**Review by yogeshbdeshpande — APPROVED** [review]

**Pull request merged**